rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,658 @@
+    ) -> Option<String> {
+        // https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingJava.html
+        if let Some(is_truncated) = list_obj_resp.is_truncated {
+            if is_truncated {
+                list_obj_resp.next_continuation_token.clone()
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+// experimental stream implementation of the list object endpoint
+struct ListObjectIdentifierFuture(
+    Pin<
+        Box<
+            dyn Future<Output = Result<(Option<Vec<ObjectIdentifier>>, Option<String>), ()>> + Send,
+        >,
+    >,
+);
+
+impl Future for ListObjectIdentifierFuture {
+    type Output = Result<(Option<Vec<ObjectIdentifier>>, Option<String>), ()>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        self.get_mut().0.as_mut().poll(cx)
+    }
+}
+
+impl ListObjectIdentifierFuture {
+    fn new(s3_client: S3Client, bucket: String, continuation_token: Option<String>) -> Self {
+        Self(Box::pin(async move {
+            let req = ListObjectsV2Request {
+                bucket: bucket.clone(),
+                continuation_token: continuation_token.clone(),
+                max_keys: Some(10),
+                ..Default::default()
+            };
+
+            let list_obj_resp = s3_client.list_objects_v2(req).await.map_err(|_| ())?;
+
+            let next_continuation_token = {
+                if let Some(is_truncated) = list_obj_resp.is_truncated {
+                    if is_truncated {
+                        list_obj_resp.next_continuation_token
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            };
+
+            let identifiers = if let Some(objects) = &list_obj_resp.contents {
+                let keys = objects
+                    .into_iter()
+                    .filter_map(|obj| obj.key.clone())
+                    .map(|key| ObjectIdentifier {
+                        key,
+                        ..Default::default()
+                    })
+                    .collect();
+                Some(keys)
+            } else {
+                None
+            };
+
+            Ok((identifiers, next_continuation_token))
+        }))
+    }
+}
+
+pub struct ListObjectsStream {
+    s3_client: S3Client,
+    bucket: String,
+    list_object_ids_future: Option<ListObjectIdentifierFuture>,
+}
+
+impl ListObjectsStream {
+    fn new(s3_client: S3Client, bucket: String) -> Self {
+        Self {
+            s3_client: s3_client.clone(),
+            bucket: bucket.clone(),
+            list_object_ids_future: Some(ListObjectIdentifierFuture::new(s3_client, bucket, None)),
+        }
+    }
+
+    fn poll_object_identifiers(&mut self, cx: &mut Context) -> Poll<Option<Vec<ObjectIdentifier>>> {
+        if let Some(ref mut fut) = self.list_object_ids_future {
```

Comment:
nit: this is shifting everything to the right. One way to avoid that is:

```rust
let mut fut = if let Some(ref mut fut) = self.list_object_ids_future {
    fut
} else {
    return Poll::Ready(None);
}
// ...
```

But it is debatable whether this is better or not to be honest so I'll leave this as a mere suggestion.